PSET: Fix blinded coinjoins with 3 or more parties

achow101 opened the pull request:
In order for blinded coinjoins with 3 or more parties to work, some fields need to be added, amounts cannot be removed, and some bugs need fixing.

First and foremost is to no longer remove amounts after blinding. Due to a miscommunication, I had believed that part of the goal of PSET was to hide semi private information (such as output amounts) from other parties in the transaction. However this causes the combiner to fail because the unique ID is dependent on those amounts and their commitments. If multiple parties had blinded just their own outputs, then the resulting PSETs would not combine because the amounts had be removed and so the unique ID calculated was incorrect. In order for this combining to work, amounts must be kept after blinding and the unique ID calculation must only use the computed unblinded transaction. This change has also been made to the spec document.

Second is the addition of explicit value and asset proofs. In order to prove that the commitments commit to the given explicit value or asset, explicit value rangeproofs and explicit asset surjection proofs are added to PSET. Each blinded output must have an explicit value rangeproof and explict asset surjection proof after blinding. For issuances, explicit issuance value proofs and explicit reissuance value proofs must be added after blinding. These proofs are verified prior to signing.

Thirdly, in order to make the coinjoin workflow work, a `blind` option is added `walletprocesspsbt` so that users can tell `walletprocesspsbt` to not attempt to blind. This is important because blinding requires all UTXOs to be present, and UTXOs can only be added via a call to `walletprocesspsbt`. Error messages have been added and improved so that users who do things in the wrong order will be less likely to end up with an unusable PSET.

Lastly, a test case has been added for a 3 party coinjoin workflow.

Fixes #1037

apoelstra commented on src/blindpsbt.cpp:
```diff
@@ -98,6 +98,23 @@ bool CreateValueRangeProof(std::vector<unsigned char>& rangeproof, const uint256
+bool CreateBlindValueProof(std::vector<unsigned char>& rangeproof, const uint256& value_blinder, const CAmount amount, const secp256k1_pedersen_commitment& value_commit, const secp256k1_generator& gen)
+{
+    // Prep rangeproof
+    size_t rangeproof_len = 5134;
```
This should be reduced (I will find the right value and revisit this comment).

Similarly it would be good to assert the value of this variable after the proof is created.

apoelstra commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1649,6 +1663,16 @@ static RPCHelpMan decodepsbt()
+        }
+
+        // Blind asset proof
+        if (!output.m_blind_value_proof.empty()) {
```
copy/paste error: `m_blind_value_proof` should be `m_blind_asset_proof`

achow101 replied:
Fixed

apoelstra commented on src/blindpsbt.cpp:
```diff
@@ -115,6 +130,26 @@ bool CreateBlindValueProof(std::vector<unsigned char>& rangeproof, const uint256
+    if (secp256k1_rangeproof_verify(secp256k1_blind_context, &min_value, &max_value, &value_commit, proof.data(), proof.size(), /* extra_commit */ nullptr, /* extra_commit_len */ 0, &gen) == 0) {
+        return false;
+    }
+    return min_value == (uint64_t)value;
```
Let me double-check on this but I believe you also need to check that `min_value == max_value`.

apoelstra commented:
This looks great! I especially like the new argument to `GetUnsignedTx` which turned out to be a very low-diff way to get the two versions of the transaction.

Aside from the nits, my only concern is about the `blind=false` option to `walletprocesspsbt`. From private conversation with you, I understand that the purpose of this is so that we don't create asset surjection proofs before all the input data has been provided (if we did it using only our own inputs, this would reduce privacy). But wouldn't it work just as well for `walletprocesspsbt` to analyze the transaction and only blind if it can find asset IDs for every input? Then the onus is no longer on the user to provide this option.

As a followup, it would probably be useful to add another field to the `walletprocesspsbt` output, aside from just `complete`, which indicates the current state of the PSET (missing input data, not yet blinded, will not be blinded, blinded, signed/complete). It would be fine to do this in a separate PR or even as part of a later Elements release.

achow101 commented:
> Aside from the nits, my only concern is about the `blind=false` option to `walletprocesspsbt`. From private conversation with you, I understand that the purpose of this is so that we don't create asset surjection proofs before all the input data has been provided (if we did it using only our own inputs, this would reduce privacy). But wouldn't it work just as well for `walletprocesspsbt` to analyze the transaction and only blind if it can find asset IDs for every input? Then the onus is no longer on the user to provide this option.

It was simpler than I had anticipated to be "smart" about this, so I have removed `blind` and implemented the "smart" method.

> As a followup, it would probably be useful to add another field to the `walletprocesspsbt` output, aside from just `complete`, which indicates the current state of the PSET (missing input data, not yet blinded, will not be blinded, blinded, signed/complete). It would be fine to do this in a separate PR or even as part of a later Elements release.

This is what `analyzepsbt` is for (I don't remember if it works currently).

apoelstra commented:
>You areIt was simpler than I had anticipated to be "smart" about this, so I have removed blind and implemented the "smart" method.

lol, on a roll with these small-diff solutions.

code review ACK. Let me do some more testing and I will merge this.

apoelstra commented on test/functional/rpc_psbt.py:
```diff
@@ -637,6 +638,65 @@ def run_ca_tests(self):
+        # Combine updated
+        comb_psbt1 = self.nodes[0].combinepsbt([up_psbt1, up_psbt2, up_psbt3])
+        # 1 and 2 blind
+        blind_psbt1 = w1.walletprocesspsbt(psbt=comb_psbt1, sign=False)["psbt"]
```
If you double-up this line, e.g. by writing
```
         blind_psbt1 = w1.walletprocesspsbt(psbt=blind_psbt1, sign=False)["psbt"]
```
after it, you will get a failure of the form `Transaction has blind inputs belonging to this blinder but does not have outputs to blind`.

The reason is that at the beginning of `BlindPSBT` in blindpsbt.cpp, we check "is everything done" by comparing the number of fully blinded outputs with the number of outputs -- but we do this **before** filtering out outputs that aren't ours. Then when we do the actual blinding, skipping all the outputs that (a) weren't ours, and (b) were already blinded, so we skip all of them. Finally at the bottom of the function we check whether we had blinded inputs (yes) and whether we blinded no outputs (also yes) and then throw this error.

Can you add this extra test line, perhaps with a comment explaining its purpose (it should be a no-op and not throw any errors), and change the logic in blindpsbt.cpp?

achow101 replied:
Fixed, with test.

apoelstra commented:
I have a couple more ntis but I'll address them in a followup PR.
